joern-cli/frontends/jimple2cpg/src/main/scala/io/joern/jimple2cpg/util/ProgramHandlingUtil.scala

```diff
@@ -1,158 +1,137 @@
 
-  logger.debug(s"Using temporary folder at $TEMP_DIR")
-
-  /** Returns the temporary directory used to unpack and analyze projects in. This allows us to lazily create the
-    * unpacking directory.
-    * @return
-    *   the path pointing to the unpacking directory.
-    */
-  def getUnpackingDir: Path =
-    TEMP_DIR match {
-      case None =>
-        val p = Files.createTempDirectory("joern-")
-        TEMP_DIR = Some(p)
-        p
-      case Some(dir) => dir
-    }
-
-  /** Inspects class files and moves them to the temp directory based on their package path.
-    *
-    * @param files
-    *   the class files to move.
-    * @return
-    *   the list of class files at their new locations.
-    */
-  def moveClassFiles(files: List[String]): List[String] = {
-    var destPath: Option[String] = None
+    def this(file: File) = this(Left(file))
+    def this(entry: ZipEntry) = this(Right(entry))
+    private def file: File            = entry.fold(identity, e => File(e.getName))
+    def name: String                  = file.name
+    def fullExtension: Option[String] = file.extension(includeAll = true)
+    def extension: Option[String]     = file.extension
+    def isDirectory: Boolean          = entry.fold(_.isDirectory, _.isDirectory)
+    def maybeRegularFile(): Boolean   = entry.fold(_.isRegularFile, !_.isDirectory)
+  }
 
-    sealed class ClassPathVisitor extends ClassVisitor(Opcodes.ASM9) {
-      override def visit(
-        version: Int,
-        access: Int,
-        name: String,
-        signature: String,
-        superName: String,
-        interfaces: Array[String]
-      ): Unit = {
-        destPath = Some(getUnpackingDir.toAbsolutePath.toString + File.separator + name + ".class")
-      }
+  private def unfoldArchives[A](src: File, emitOrUnpack: File => Either[A, List[File]]): IterableOnce[A] = {
+    emitOrUnpack(src) match {
+      case Left(a)             => Seq(a)
+      case Right(disposeFiles) => disposeFiles.flatMap(x => unfoldArchives(x, emitOrUnpack))
     }
+  }
+  private def extractClassesToTmp(
+    src: File,
+    tmpDir: File,
+    isArchive: Entry => Boolean,
+    isSource: Entry => Boolean
+  ): IterableOnce[ClassFile] = {
 
-    files.flatMap { f =>
-      Using.resource(new FileInputStream(f)) { fis =>
-        val cr          = new ClassReader(fis)
-        val rootVisitor = new ClassPathVisitor()
-        cr.accept(rootVisitor, SKIP_CODE)
+    def shouldExtract(e: Entry) = e.maybeRegularFile() && (isArchive(e) || isSource(e))
+    unfoldArchives(
+      src,
+      {
+        case f if isSource(Entry(f)) =>
+          Left(ClassFile(f))
+        case f if f.isDirectory() =>
+          val files = f.listRecursively.filterNot(_.isDirectory).toList
+          Right(files)
+        case f if isArchive(Entry(f)) =>
+          val xTmp = File.newTemporaryDirectory("extract-archive-", parent = Some(tmpDir))
+          val unzipDirs = Try(f.unzipTo(xTmp, e => shouldExtract(Entry(f)))) match {
```

Comment:
One thing @fabsx00 pointed out to me is that we should be aware of zip-slip vulnerabilities, e.g. if a zip entry contains a `../` path to get you to write outside the destination output. So perhaps find a way to normalize or filter out these kinds of paths that contain `..` in them